### PR TITLE
add shell integration setting

### LINKED ISSUE #139398: Shell integration: Write bash shell integration script
Parent issue https://github.com/microsoft/vscode/issues/133084

### PR DESCRIPTION
**Pull request by meganrogge**
fixes #139398

### CODE REVIEW

File: src/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+precmd() {
+    local STATUS="$?"
+    if [ -z "${IN_COMMAND_EXECUTION-}" ]; then
+        # ctrl c was = to ""
+        command_output_start
+    fi
+
+    command_complete "$STATUS"
+
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh:47]: works (saves commands and cwds), but messes w the prompt

File: src/vs/workbench/contrib/terminal/common/terminalConfiguration.ts [hunk 2]
```diff
@@ -514,7 +514,12 @@ const terminalConfiguration: IConfigurationNode = {
 			default: {
 				'Terminate batch job (Y/N)': 'Y\r'
 			}
-		}
+		},
+		[TerminalSettingId.EnableShellIntegration]: {
+			description: localize('terminal.integrated.enableShellIntegration', "For supported shells, use shell integration (zsh on mac/linux, pwsh on windows)"),
+			type: 'boolean',
+			default: true
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminal/common/terminalConfiguration.ts:521]: false

File: src/vs/workbench/contrib/terminal/browser/xterm/shellIntegrationAddon.ts [hunk 3]
```diff
@@ -114,6 +114,8 @@ export class ShellIntegrationAddon extends Disposable implements IShellIntegrati
 				return false;
 		}
 		const value = exitCode || type;
+		console.log(command, value);
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminal/browser/xterm/shellIntegrationAddon.ts:117]: remove

File: src/vs/workbench/contrib/terminal/browser/xterm/cognisantCommandTrackerAddon.ts [hunk 4]
```diff
@@ -48,6 +48,7 @@ export class CognisantCommandTrackerAddon extends CommandTrackerAddon {
 		if (!this._terminal) {
 			return;
 		}
+		console.log('type', event.type);
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminal/browser/xterm/cognisantCommandTrackerAddon.ts:51]: remove

File: src/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        update_prompt
+    fi
+}
+preexec() {
+    PS1="$PRIOR_PROMPT"
+    IN_COMMAND_EXECUTION="1"
+    command_output_start
+}
+update_prompt
+PROMPT_COMMAND=${PROMPT_COMMAND:+"$PROMPT_COMMAND; "}'precmd'
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh:60]: Does this work with how I use `PROMPT_COMMAND`?

https://github.com/Tyriar/dotfiles/blob/a9fd490615823056ae0ff8bd9b1b5e95cd185a0b/data/bash/.bash_prompt#L1-L22

File: src/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+set_shell_integration_enabled() {
+    printf "\033]133;E\007"
+}
+
+update_prompt() {
+    PRIOR_PROMPT="$PS1"
+    IN_COMMAND_EXECUTION=""
+    local PREFIX=""
+    if [[ $PS1 == *"$(prompt_start)"* ]]; then
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh:31]: It's not clear what this is doing? If `update_prompt` is called multiple times I'm not sure why?

File: src/vs/workbench/contrib/terminal/browser/media/shellIntegration-zsh.sh [hunk 7]
```diff
@@ -0,0 +1,63 @@
+    IN_COMMAND_EXECUTION="1"
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/media/shellIntegration-zsh.sh:1]: We should remove this indentation

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 8]
```diff
@@ -128,6 +128,7 @@ export class TerminalInstance extends Disposable implements ITerminalInstance {
 	private _latestXtermWriteData: number = 0;
 	private _latestXtermParseData: number = 0;
 	private _isExiting: boolean;
+	private _enableShellIntegration: boolean = false;
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:131]: I think you said the 133;E sequence was custom in which case we would want to use our own OSC code for. But I've been thinking we shouldn't need that at all, we could infer shell integration is on if needed by querying the capabilities it turns on.

**Pull request merged**